relay-server/src/actors/project_cache.rs

```diff
@@ -1091,4 +1105,60 @@ mod tests {
             assert!(buffer_rx.try_recv().is_err())
         }
     }
+
+    #[tokio::test]
+    async fn test_eviction() {
+        let num_permits = 5;
+        let buffer_guard: Arc<_> = BufferGuard::new(num_permits).into();
+        let services = mocked_services();
+        let (state_tx, _) = mpsc::unbounded_channel();
+        let (buffer_tx, _) = mpsc::unbounded_channel();
+
+        // Projects should be expired after 2 seconds.
+        let config: Arc<_> = Config::from_json_value(serde_json::json!({
+            "cache": {
+              "project_expiry": 1
+            }
+        }))
+        .unwrap()
+        .into();
+
+        let (mut broker, _) = project_cache_broker_setup(
+            config.clone(),
+            services.clone(),
+            buffer_guard,
+            state_tx,
+            buffer_tx,
+        )
+        .await;
+
+        let key1 = ProjectKey::parse("e12d836b15bb49d7bbf99e64295d995b").unwrap();
+        let project1 = Project::new(key1, config.clone());
+        let key2 = ProjectKey::parse("eeed836b15bb49d7bbf99e64295d9bbb").unwrap();
+        let project2 = Project::new(key2, config);
+
+        broker.projects.insert(key1, project1);
+        broker.projects.insert(key2, project2);
+
+        for _ in 0..20 {
+            let envelope = ManagedEnvelope::untracked(
+                empty_envelope(),
+                services.outcome_aggregator.clone(),
+                services.test_store.clone(),
+            );
+            if let hashbrown::hash_map::Entry::Occupied(mut e) = broker.projects.entry(key2) {
+                e.get_mut()
+                    .check_envelope(envelope, services.outcome_aggregator.clone())
+                    .unwrap();
+            }
+
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // One of the project will be removed.
+        broker.evict_stale_project_caches();
+        // Project 1 did not receive any envelopes, and should be removed now.
+        assert_eq!(broker.projects.len(), 1);
+        assert!(broker.projects.get(&key2).is_some());
```

Comment:
To make the difference to the old implementation more obvious, can we assert here that `project2` has an outdated project state i.e. `last_updated_at < X`. And maybe check that `project2.expiry_state() == ExpiryState::Expired`, to make sure that these envelopes will not be processed but rather enqueued.